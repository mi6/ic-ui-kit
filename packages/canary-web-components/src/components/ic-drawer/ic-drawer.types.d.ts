export interface IcDrawerExpandedDetail {
  expanded: boolean;
}
