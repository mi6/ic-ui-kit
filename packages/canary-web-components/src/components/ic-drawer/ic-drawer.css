@import "../../../dist/core/normalize.css";

/**
  @prop --ic-z-index-drawer: z-index of drawer.
  @prop --ic-drawer-width: the width of the expanded left or right drawer.
  @prop --ic-drawer-height: the width of the expanded top or bottom drawer.
*/

:host(.ic-drawer) {
  position: fixed;
  display: flex;
  top: 0;
  left: 0;
  z-index: var(--ic-z-index-drawer);

  --vertical-drawer-line-clamp: 0;
}

/* Collapsed state */

:host(.ic-drawer.collapsed) {
  flex: 0 0 56px;
}

:host(.ic-drawer) .collapsed-drawer {
  position: fixed;
  display: flex;
  justify-content: center;
  background-color: var(--ic-architectural-white);
  box-shadow: var(--ic-elevation-overlay);
}

:host(.ic-drawer.ic-left-position) .collapsed-drawer,
:host(.ic-drawer.ic-right-position) .collapsed-drawer {
  height: 100%;
}

:host(.ic-drawer.ic-top-position) .collapsed-drawer,
:host(.ic-drawer.ic-bottom-position) .collapsed-drawer {
  width: 100%;
  align-items: center;
}

:host(.ic-drawer.ic-left-position) .collapsed-drawer .chevron-btn,
:host(.ic-drawer.ic-right-position) .collapsed-drawer .chevron-btn {
  margin: var(--ic-space-md) var(--ic-space-sm) 0;
}

:host(.ic-drawer.ic-top-position) .collapsed-drawer .chevron-btn,
:host(.ic-drawer.ic-bottom-position) .collapsed-drawer .chevron-btn {
  display: flex;
  flex: 1 1 0;
  justify-content: center;
  margin: var(--ic-space-xs) 0;
}

/* Expanded state */

:host(.ic-drawer.expanded) {
  width: 100vw;
  height: 100vh;
}

:host(.ic-drawer) .ic-drawer-overlay {
  z-index: -1;
  flex: auto;
  background-color: rgb(0 0 0 / 20%);
}

:host(.ic-drawer) .ic-drawer-panel {
  display: flex;
  position: fixed;
  background-color: var(--ic-architectural-white);
  box-shadow: var(--ic-elevation-overlay);
}

:host(.ic-drawer.ic-top-position) .ic-drawer-panel,
:host(.ic-drawer.ic-bottom-position) .ic-drawer-panel {
  width: 100%;
}

:host(.ic-drawer.ic-right-position.expanded) .ic-drawer-panel.small,
:host(.ic-drawer.ic-left-position.expanded) .ic-drawer-panel.small {
  width: var(--ic-drawer-width, calc(var(--ic-space-sm) * 25));
}

:host(.ic-drawer.ic-right-position.expanded) .ic-drawer-panel.medium,
:host(.ic-drawer.ic-left-position.expanded) .ic-drawer-panel.medium {
  width: var(--ic-drawer-width, calc(var(--ic-space-sm) * 37.5));
}

:host(.ic-drawer.ic-right-position.expanded) .ic-drawer-panel.large,
:host(.ic-drawer.ic-left-position.expanded) .ic-drawer-panel.large {
  width: var(--ic-drawer-width, calc(var(--ic-space-sm) * 50));
}

:host(.ic-drawer.ic-top-position.expanded) .ic-drawer-panel.small,
:host(.ic-drawer.ic-bottom-position.expanded) .ic-drawer-panel.small {
  height: var(--ic-drawer-height, calc(var(--ic-space-sm) * 25));
}

:host(.ic-drawer.ic-top-position.expanded) .ic-drawer-panel.medium,
:host(.ic-drawer.ic-bottom-position.expanded) .ic-drawer-panel.medium {
  height: var(--ic-drawer-height, calc(var(--ic-space-sm) * 37.5));
}

:host(.ic-drawer.ic-top-position.expanded) .ic-drawer-panel.large,
:host(.ic-drawer.ic-bottom-position.expanded) .ic-drawer-panel.large {
  height: var(--ic-drawer-height, calc(var(--ic-space-sm) * 50));
}

:host(.ic-drawer.ic-right-position),
:host(.ic-drawer.ic-left-position) {
  top: 0;
  bottom: 0;
}

:host(.ic-drawer.ic-right-position.expanded) .ic-drawer-panel,
:host(.ic-drawer.ic-left-position.expanded) .ic-drawer-panel {
  height: 100%;
}

:host(.ic-drawer) .inner-drawer-panel {
  display: flex;
  flex-flow: column;
  flex: 1 1 0;

  --section-container-margin: 0;
}

:host(.ic-drawer.ic-top-position),
:host(.ic-drawer.ic-bottom-position) {
  left: 0;
  right: 0;
}

:host(.ic-drawer.ic-right-position.expanded) {
  transition: right var(--ic-easing-transition-slow);
  right: 0;
}

:host(.ic-drawer.ic-right-position) .collapsed-drawer,
:host(.ic-drawer.ic-right-position) .ic-drawer-panel {
  right: 0;
}

:host(.ic-drawer.ic-left-position) .collapsed-drawer,
:host(.ic-drawer.ic-left-position) .ic-drawer-panel {
  left: 0;
}

:host(.ic-drawer.ic-top-position) .collapsed-drawer,
:host(.ic-drawer.ic-top-position) .ic-drawer-panel {
  top: 0;
}

:host(.ic-drawer.ic-bottom-position) .collapsed-drawer,
:host(.ic-drawer.ic-bottom-position) .ic-drawer-panel {
  bottom: 0;
}

/* Drawer panel styling */

:host(.ic-drawer) .drawer-header {
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: var(--ic-space-sm) var(--ic-space-xs);
  border-bottom: var(--ic-border-light);
}

:host(.ic-drawer) .heading-area {
  display: flex;
  align-items: center;
  gap: var(--ic-space-xs);
}

:host(.ic-drawer) .heading-area.no-arrow {
  padding-left: var(--ic-space-xs);
}

:host(.ic-drawer) .main-content {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  width: 100%;
  height: calc(100% - var(--ic-space-xl));
  margin-bottom: var(--ic-space-xs);

  --section-container-margin: 0;
}

:host(.ic-drawer) .message-area {
  overflow-y: auto;
  flex: 1 1 0;
  padding-top: var(--ic-space-md);
}

:host(.ic-drawer) .message-area.message-area-padding {
  margin: var(--ic-space-md) var(--ic-space-md) 0;
}

:host(.ic-drawer) .body-text.message-area-padding {
  padding: 0 var(--ic-space-md);
}

:host(.ic-drawer) .message-area .body-text {
  height: 100%;

  /* overflow-y: auto; */

  /* box-shadow: 0 4px 16px rgb(0 0 0 / 25%) */
  background:
  /* Shadow Cover TOP */ linear-gradient(
        white 30%,
        rgb(255 255 255 / 0%)
      )
      center top,
    /* Shadow Cover BOTTOM */ linear-gradient(rgb(255 255 255 / 0%), white 70%)
      center bottom,
    /* Shadow TOP */
      linear-gradient(
        to bottom,
        rgb(0 0 0 / 10%) 0%,
        rgb(0 0 0 / 5%) 50%,
        rgb(0 0 0 / 0%)
      )
      center top,
    /* Shadow BOTTOM */
      linear-gradient(
        to top,
        rgb(0 0 0 / 10%) 0%,
        rgb(0 0 0 / 5%) 40%,
        rgb(0 0 0 / 5%) 70%,
        rgb(0 0 0 / 0%)
      )
      center bottom;
  background-repeat: no-repeat;
  background-size: 100% 12px, 100% 12px, 100% 12px, 100% 12px;
  background-attachment: local, local, scroll, scroll;
  white-space: break-spaces;
}

:host(.ic-drawer) .action-area {
  display: flex;
  justify-content: flex-end;
  margin-top: var(--ic-space-md);
  padding: 0 var(--ic-space-md);
  gap: var(--ic-space-md);
}

/* box-shadow: var(--ic-elevation-inset); */

/* :host(.ic-drawer) .action-area.main-content-overflow {
  box-shadow: var(--ic-elevation-overlay);
} */

/* Chevron styling */
:host(.ic-drawer) .chevron-btn::part(button) {
  padding: 0;
}

/* :host(.ic-drawer) .chevron-btn {
  --height: calc(var(--ic-space-md) * 2);--min-width: calc(var(--ic-space-md) * 2);

} */

:host(.ic-drawer.ic-left-position.expanded) .chevron-btn svg,
:host(.ic-drawer.ic-right-position.collapsed) .chevron-btn svg {
  transform: rotate(180deg);
}

:host(.ic-drawer.ic-top-position.expanded) .chevron-btn svg,
:host(.ic-drawer.ic-bottom-position.collapsed) .chevron-btn svg {
  transform: rotate(270deg);
}

:host(.ic-drawer.ic-top-position.collapsed) .chevron-btn svg,
:host(.ic-drawer.ic-bottom-position.expanded) .chevron-btn svg {
  transform: rotate(90deg);
}

.drawer-expanded-body {
  overflow: hidden !important;
}

/* High contrast mode */
@media (forced-colors: active) {
  :host(.ic-drawer.ic-left-position) .collapsed-drawer,
  :host(.ic-drawer.ic-left-position) .ic-drawer-panel {
    border-right: solid 1px transparent;
  }

  :host(.ic-drawer.ic-right-position) .collapsed-drawer,
  :host(.ic-drawer.ic-right-position) .ic-drawer-panel {
    border-left: solid 1px transparent;
  }

  :host(.ic-drawer.ic-top-position) .collapsed-drawer,
  :host(.ic-drawer.ic-top-position) .ic-drawer-panel {
    border-bottom: solid 1px transparent;
  }

  :host(.ic-drawer.ic-bottom-position) .collapsed-drawer,
  :host(.ic-drawer.ic-bottom-position) .ic-drawer-panel {
    border-top: solid 1px transparent;
  }
}
