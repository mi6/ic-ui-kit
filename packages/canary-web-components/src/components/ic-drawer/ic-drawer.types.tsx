export interface IcDrawerExpandedDetail {
  expanded: boolean;
}

export type IcDrawerBoundary = "viewport" | "parent";
