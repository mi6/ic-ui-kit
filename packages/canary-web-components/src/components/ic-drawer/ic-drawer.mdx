import { Meta, Stories } from "@storybook/blocks";
import * as IcDrawerStories from "./ic-drawer.stories";

<Meta of={IcDrawerStories} />

<Stories />
