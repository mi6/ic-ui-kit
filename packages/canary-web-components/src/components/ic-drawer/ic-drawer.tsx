import {
  Component,
  Host,
  h,
  Prop,
  Element,
  Event,
  EventEmitter,
  Listen,
} from "@stencil/core";

import { IcPosition } from "@ukic/web-components/dist/types/utils/types";

import chevronIcon from "../../assets/chevron-icon.svg";

import closeIcon from "../../assets/close-icon.svg";

import { isSlotUsed } from "../../utils/helpers";
import { IcDrawerExpandedDetail } from "./ic-drawer.types";
import { isPropDefined } from "../../utils/helpers";

/**
 * @slot heading - Content will be rendered in the title area, in place of the heading.
 * @slot heading-adornment - Content will be rendered in the title area, to the left of the heading.
 * @slot message - Content will be rendered below the title area, in place of the message.
 * @slot actions - Content will be rendered in the action area, adjacent to the title area.
 */

@Component({
  tag: "ic-drawer",
  styleUrl: "ic-drawer.css",
  shadow: true,
})
export class Drawer {
  private DEFAULT_CLOSE_BUTTON_ARIA_LABEL = "Close drawer";
  private IC_ACCORDION: string = "IC-ACCORDION";
  private IC_ACCORDION_GROUP: string = "IC-ACCORDION-GROUP";
  private IC_TEXT_FIELD: string = "IC-TEXT-FIELD";

  private focusedElementIndex = 0;
  private interactiveElementList: HTMLElement[];

  @Element() el: HTMLIcDrawerElement;

  /**
   * The aria label applied to the drawer. This is required when the heading slot is used.
   */
  @Prop() ariaLabel: string = "";

  /**
   * The aria label of the chevron button when trigger is set to "arrow". The default aria label is "Open drawer" / "Close drawer".
   */
  @Prop() chevronButtonAriaLabel?: string = "";

  /**
   * The aria label of the close button when the show-close-button attribute (web-components) / showCloseButton prop (React) is set to `true`. The default aria label is "Open drawer" / "Close drawer".
   */
  @Prop() closeButtonAriaLabel: string = "Close drawer";

  /**
   * If set to `false`, the drawer will not close when the backdrop is clicked.
   */
  @Prop() closeOnBackdropClick?: boolean = true;

  /**
   * If set to `true`, the drawer will display in an expanded state.
   */
  @Prop({ reflect: true, mutable: true }) expanded: boolean = false;

  /**
   * The heading to render in the drawer.
   */
  @Prop() heading?: string;

  /**
   * The main body text in the drawer.
   */
  @Prop() message?: string;

  /**
   * The position of the drawer.
   */
  @Prop() position: IcPosition = "right";

  /**
   * If set to `true`, an X (close) button will be displayed in the drawer.
   */
  @Prop() showCloseButton: boolean = false;

  /**
   * The size of the expanded drawer.
   */
  @Prop() size?: "small" | "medium" | "large" = "medium";

  /**
   * The method in which the drawer is expanded.
   */
  @Prop() trigger: "arrow" | "controlled" = "arrow";

  /**
   * The variant of the drawer. TODO - figure out how this works
   */
  @Prop() variant: "temporary" | "persistent" = "temporary";

  /**
   * Emitted when the drawer is collapsed and expanded.
   */
  @Event() icDrawerExpanded: EventEmitter<IcDrawerExpandedDetail>;

  @Listen("keydown", { target: "document" })
  handleKeyboard(ev: KeyboardEvent): void {
    if (this.expanded) {
      switch (ev.key) {
        case "Tab":
          ev.preventDefault();
          this.focusNextInteractiveElement(ev.shiftKey);
          break;
        case "Escape":
          this.setDrawerExpanded();
          break;
      }
    }
  }

  componentDidLoad(): void {
    this.updateInteractiveElements();
  }

  componentDidUpdate(): void {
    this.updateInteractiveElements();
  }

  private getFocusedElementIndex = () => {
    for (let i = 0; i < this.interactiveElementList.length; i++) {
      if (
        (this.interactiveElementList[i] as HTMLElement) ===
        (this.el.shadowRoot.activeElement || document.activeElement)
      ) {
        this.focusedElementIndex = i;
      }
    }
  };

  private getInteractiveElements = () => {
    this.interactiveElementList = Array.from(
      this.el.shadowRoot.querySelectorAll("ic-button")
    );

    const slottedInteractiveElements = Array.from(
      this.el.querySelectorAll(
        `a[href], button, input:not(.ic-input), textarea, select, details, [tabindex]:not([tabindex="-1"]), 
          ic-button, ic-checkbox, ic-select, ic-search-bar, ic-tab-group, ic-radio-group, 
          ic-back-to-top, ic-breadcrumb, ic-chip[dismissible="true"], ic-footer-link, ic-link, ic-navigation-button, 
          ic-navigation-item, ic-switch, ic-text-field, ic-accordion-group, ic-accordion`
      )
    );

    const messageArea = this.el.shadowRoot.querySelector(
      ".message-area"
    ) as HTMLElement;

    const isScrollable = (el: HTMLElement): boolean =>
      el.scrollHeight > el.clientHeight;

    // Check if the message area is scrollable and include it as an interactive element
    if (messageArea && isScrollable(messageArea)) {
      this.interactiveElementList.push(messageArea);
    }

    this.interactiveElementList = [
      ...this.interactiveElementList,
      ...(slottedInteractiveElements as HTMLElement[]),
    ];
  };

  // When trigger is controlled, focus should be set to the first interactive element on drawer expansion

  private getNextFocusEl = (focusedElementIndex: number) =>
    this.interactiveElementList[focusedElementIndex];

  private focusNextInteractiveElement = (shiftKey: boolean) => {
    this.getFocusedElementIndex();
    this.setFocusIndexBasedOnShiftKey(shiftKey);
    this.loopNextFocusIndexIfLastElement();

    let nextFocusEl = this.getNextFocusEl(this.focusedElementIndex);

    const isHidden =
      getComputedStyle(nextFocusEl).visibility === "hidden" ||
      (nextFocusEl.tagName === this.IC_ACCORDION_GROUP &&
        nextFocusEl.hasAttribute("single-expansion"));

    if (nextFocusEl.tagName === this.IC_TEXT_FIELD) {
      (nextFocusEl as HTMLIcTextFieldElement).setFocus();
    } else {
      if (isHidden) {
        this.setFocusIndexBasedOnShiftKey(shiftKey);
        this.loopNextFocusIndexIfLastElement();

        nextFocusEl = this.getNextFocusEl(this.focusedElementIndex);
      }
      if (nextFocusEl.tagName === this.IC_ACCORDION_GROUP) {
        (nextFocusEl as HTMLIcAccordionGroupElement).setFocus();
      } else if (nextFocusEl.tagName === this.IC_ACCORDION) {
        (nextFocusEl as HTMLIcAccordionElement).setFocus();
      } else {
        (nextFocusEl as HTMLElement).focus();
      }
    }
  };

  private loopNextFocusIndexIfLastElement() {
    if (this.focusedElementIndex > this.interactiveElementList.length - 1)
      this.focusedElementIndex = 0;
    else if (this.focusedElementIndex < 0) {
      this.focusedElementIndex = this.interactiveElementList.length - 1;
    }
  }

  private setFocusIndexBasedOnShiftKey(shiftKey: boolean) {
    if (shiftKey) {
      this.focusedElementIndex -= 1;
    } else {
      this.focusedElementIndex += 1;
    }
  }

  private updateInteractiveElements = () => {
    if (this.expanded) {
      this.getInteractiveElements();
    }
  };

  private setDrawerExpanded = (): void => {
    this.expanded = !this.expanded;
    this.getInteractiveElements();
    if (this.expanded && this.interactiveElementList.length > 0) {
      console.log(this.interactiveElementList);
      if (this.interactiveElementList[0].tagName === "IC-BUTTON") {
        (this.interactiveElementList[0] as HTMLIcButtonElement).setFocus();
        console.log(
          (
            this.interactiveElementList[0] as HTMLIcButtonElement
          ).shadowRoot.querySelector("button")
        );
      } else {
        this.interactiveElementList[1].focus();
      }
    }

    this.icDrawerExpanded.emit({ expanded: this.expanded });
  };

  private onBackdropClick = (e: { stopPropagation: () => void }) => {
    if (this.closeOnBackdropClick) {
      this.setDrawerExpanded();
    }
    e.stopPropagation();
  };

  render() {
    const {
      chevronButtonAriaLabel,
      closeButtonAriaLabel,
      expanded,
      heading,
      message,
      position,
      showCloseButton,
      size,
      trigger,
    } = this;

    const chevronButton = (
      <ic-button
        class="chevron-btn"
        size="small"
        variant="icon"
        innerHTML={chevronIcon}
        aria-label={`${
          isPropDefined(chevronButtonAriaLabel)
            ? chevronButtonAriaLabel
            : expanded
            ? this.DEFAULT_CLOSE_BUTTON_ARIA_LABEL
            : "Open drawer"
        }`}
        onClick={this.setDrawerExpanded}
      ></ic-button>
    );

    return (
      <Host
        class={{
          ["ic-drawer"]: true,
          [`ic-${position}-position`]: true,
          ["expanded"]: expanded,
          ["collapsed"]: !expanded,
        }}
        aria-expanded={expanded}
      >
        {!expanded && trigger === "arrow" && (
          <div class="collapsed-drawer">{chevronButton}</div>
        )}
        {expanded && (
          <div
            class={{ ["ic-drawer-overlay"]: true }}
            onClick={this.onBackdropClick}
          ></div>
        )}
        <div
          class={{ ["ic-drawer-panel"]: true, [`${size}`]: true }}
          {...(expanded && { role: "dialog" })}
          {...(!isSlotUsed(this.el, heading)
            ? {
                "aria-labelledby":
                  this.ariaLabel !== "" ? this.ariaLabel : "ic-drawer-heading",
              }
            : {})}
          {...(isSlotUsed(this.el, heading)
            ? { "aria-label": this.ariaLabel }
            : {})}
        >
          {expanded && (
            <ic-section-container
              aligned="full-width"
              fullHeight
              class="inner-drawer-panel"
            >
              <a id="drawer-content"></a>
              <div class="drawer-header">
                <div
                  class={{
                    ["heading-area"]: true,
                    ["no-arrow"]: trigger !== "arrow",
                  }}
                >
                  {trigger === "arrow" && chevronButton}
                  {isSlotUsed(this.el, "heading-adornment") && (
                    <slot name="heading-adornment" />
                  )}
                  {isSlotUsed(this.el, "heading") ? (
                    <slot name="heading" />
                  ) : (
                    <ic-typography
                      class="drawer-heading"
                      id="ic-drawer-heading"
                      variant="h4"
                    >
                      {heading}
                    </ic-typography>
                  )}
                </div>
                {showCloseButton && trigger === "controlled" && (
                  <ic-button
                    className="close-btn"
                    variant="icon"
                    appearance="dark"
                    onClick={this.setDrawerExpanded}
                    innerHTML={closeIcon}
                    aria-label={
                      isPropDefined(closeButtonAriaLabel)
                        ? closeButtonAriaLabel
                        : this.DEFAULT_CLOSE_BUTTON_ARIA_LABEL
                    }
                  ></ic-button>
                )}
              </div>
              <ic-section-container
                aligned="full-width"
                fullHeight
                class="main-content"
              >
                {(isSlotUsed(this.el, "message") || !!message) && (
                  <div
                    class={{
                      ["message-area"]: true,
                      ["message-area-padding"]: isSlotUsed(this.el, "message"),
                    }}
                    tabindex={0}
                  >
                    {isSlotUsed(this.el, "message") ? (
                      <slot name="message" />
                    ) : (
                      <ic-typography
                        class={{
                          ["body-text"]: true,
                          ["message-area-padding"]: !isSlotUsed(
                            this.el,
                            "message"
                          ),
                        }}
                      >
                        <p>{message}</p>
                      </ic-typography>
                    )}
                  </div>
                )}
                {isSlotUsed(this.el, "actions") && (
                  <div
                    class={{
                      ["action-area"]: true,
                      ["main-content-overflow"]: true,
                    }}
                  >
                    <slot name="actions" />
                  </div>
                )}
              </ic-section-container>
            </ic-section-container>
          )}
        </div>
      </Host>
    );
  }
}
