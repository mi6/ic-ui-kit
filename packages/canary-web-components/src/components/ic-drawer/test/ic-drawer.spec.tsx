import { newSpecPage } from "@stencil/core/testing";

import { Drawer } from "../ic-drawer";
import { testKeyboardEvent as keyboardEvent } from "../../../testspec.setup";

describe("ic-drawer", () => {
  it("should render on the right", async () => {
    const page = await newSpecPage({
      components: [Drawer],
      html: `<ic-drawer heading="Menu" message="test" position="right"></ic-drawer>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render on the left", async () => {
    const page = await newSpecPage({
      components: [Drawer],
      html: `<ic-drawer heading="Menu" message="test" position="left"></ic-drawer>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render at the top of the viewport", async () => {
    const page = await newSpecPage({
      components: [Drawer],
      html: `<ic-drawer heading="Menu" message="test" position="top"></ic-drawer>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render at the bottom of the viewport", async () => {
    const page = await newSpecPage({
      components: [Drawer],
      html: `<ic-drawer heading="Menu" message="test" position="bottom"></ic-drawer>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should set expanded as false and stop other events propagating when the backdrop is clicked when closeOnBackdropClick is set to true", async () => {
    const page = await newSpecPage({
      components: [Drawer],
      html: `<ic-drawer heading="Menu" message="test"expanded="true"></ic-drawer>`,
    });

    expect(page.rootInstance.closeOnBackdropClick).toBe(true);

    const mockEvent = {
      stopPropagation: jest.fn(),
    };

    await page.rootInstance.onBackdropClick(mockEvent);

    expect(page.rootInstance.expanded).toBe(false);
  });

  it("should not set expanded as false when the backdrop is clicked when closeOnBackdropClick is set to false", async () => {
    const page = await newSpecPage({
      components: [Drawer],
      html: `<ic-drawer heading="Menu" message="test" expanded="true" close-on-backdrop-click="false"></ic-drawer>`,
    });

    const mockEvent = {
      stopPropagation: jest.fn(),
    };

    const setDrawerExpandedSpy = jest.spyOn(
      page.rootInstance,
      "setDrawerExpanded"
    );

    await page.rootInstance.onBackdropClick(mockEvent);

    expect(setDrawerExpandedSpy).toHaveBeenCalledTimes(0);
  });

  it("should call the setDrawerExpanded method correctly", async () => {
    const page = await newSpecPage({
      components: [Drawer],
      html: `<ic-drawer heading="Menu" message="test"><div slot="actions">
          <ic-button id="btn1" >Button 1</ic-button>
          <ic-button id="btn2">Button 2</ic-button>
          <ic-button id="btn3">Button 3</ic-button>
        </div>
      </ic-drawer>`,
    });
    expect(page.rootInstance.expanded).toBe(false);

    await page.rootInstance.setDrawerExpanded();

    expect(page.rootInstance.expanded).toBe(true);

    expect(page.body.style.overflow).toBe("hidden");
  });

  it("should return the correct index of the focused element", async () => {
    const page = await newSpecPage({
      components: [Drawer],
      html: `<ic-drawer heading="Menu" message="test" expanded="true">
        <div slot="actions">
          <ic-button id="btn1">Button 1</ic-button>
          <ic-button id="btn2">Button 2</ic-button>
          <ic-button id="btn3">Button 3</ic-button>
        </div>
      </ic-drawer>`,
    });

    const getFocusedElementIndexSpy = jest.spyOn(
      page.rootInstance,
      "getFocusedElementIndex"
    );

    const setFocusIndexBasedOnShiftKeySpy = jest.spyOn(
      page.rootInstance,
      "setFocusIndexBasedOnShiftKey"
    );

    const loopNextFocusIndexIfLastElementSpy = jest.spyOn(
      page.rootInstance,
      "loopNextFocusIndexIfLastElement"
    );

    const getNextFocusElSpy = jest.spyOn(page.rootInstance, "getNextFocusEl");

    // Set up the interactiveElementList
    const drawer = page.rootInstance;
    drawer.interactiveElementList = [
      page.root.querySelector("#btn1"),
      page.root.querySelector("#btn2"),
      page.root.querySelector("#btn3"),
    ];

    page.rootInstance.focusedElementIndex = 2;

    // Call the focusNextInteractiveElement function
    drawer.focusNextInteractiveElement(true);

    expect(getFocusedElementIndexSpy).toHaveBeenCalledTimes(1);

    expect(setFocusIndexBasedOnShiftKeySpy).toHaveBeenCalledTimes(1);
    expect(setFocusIndexBasedOnShiftKeySpy).toHaveBeenCalledWith(true);

    // Assert that the focusedElementIndex is now 1
    expect(page.rootInstance.focusedElementIndex).toBe(1);

    expect(loopNextFocusIndexIfLastElementSpy).toHaveBeenCalled();

    expect(getNextFocusElSpy).toHaveBeenCalledWith(1);
  });

  it("should not respond to pressing the escape key when the drawer is not expanded", async () => {
    const page = await newSpecPage({
      components: [Drawer],
      html: `<ic-drawer heading="Menu" message="test"></ic-drawer>`,
    });

    const setDrawerExpandedSpy = jest.spyOn(
      page.rootInstance,
      "setDrawerExpanded"
    );

    page.win.document.dispatchEvent(
      new KeyboardEvent("keydown", keyboardEvent("Escape"))
    );

    expect(setDrawerExpandedSpy).toHaveBeenCalledTimes(0);
  });

  it("should close when the drawer is expanded and the Escape key is pressed", async () => {
    const page = await newSpecPage({
      components: [Drawer],
      html: `<ic-drawer heading="Menu" message="test" expanded="true"></ic-drawer>`,
    });

    const setDrawerExpandedSpy = jest.spyOn(
      page.rootInstance,
      "setDrawerExpanded"
    );

    page.win.document.dispatchEvent(
      new KeyboardEvent("keydown", keyboardEvent("Escape"))
    );

    expect(setDrawerExpandedSpy).toHaveBeenCalledTimes(1);
  });

  it("should not create a focus trap when the Tab key is pressed and the drawer is not expanded", async () => {
    const page = await newSpecPage({
      components: [Drawer],
      html: `<ic-drawer heading="Menu" message="test"></ic-drawer>`,
    });

    const event = new KeyboardEvent("keydown", { key: "Tab" });

    jest.spyOn(event, "preventDefault");

    const focusNextInteractiveElementSpy = jest.spyOn(
      page.rootInstance,
      "focusNextInteractiveElement"
    );

    page.win.document.dispatchEvent(
      new KeyboardEvent("keydown", keyboardEvent("Tab"))
    );

    expect(event.preventDefault).toHaveBeenCalledTimes(0);
    expect(focusNextInteractiveElementSpy).toHaveBeenCalledTimes(0);
  });

  it("should create a focus trap when the Tab key is pressed and the drawer is expanded", async () => {
    const page = await newSpecPage({
      components: [Drawer],
      html: `<ic-drawer heading="Menu" message="test" expanded="true"></ic-drawer>`,
    });

    const event = new KeyboardEvent("keydown", { key: "Tab" });

    jest.spyOn(event, "preventDefault");

    const focusNextInteractiveElementSpy = jest.spyOn(
      page.rootInstance,
      "focusNextInteractiveElement"
    );

    page.win.document.dispatchEvent(event);

    expect(event.preventDefault).toHaveBeenCalledTimes(1);
    expect(focusNextInteractiveElementSpy).toHaveBeenCalledTimes(1);
  });

  describe("ic-drawer lifecycle methods", () => {
    it(`should call updateInteractiveElements in componentDidLoad`, async () => {
      const page = await newSpecPage({
        components: [Drawer],
        html: `<ic-drawer heading="Menu" message="test"></ic-drawer>`,
      });

      const updateInteractiveElementsSpy = jest.spyOn(
        page.rootInstance,
        "updateInteractiveElements"
      );

      page.rootInstance.componentDidLoad();

      expect(updateInteractiveElementsSpy).toHaveBeenCalledTimes(1);
    });

    it("should call updateInteractiveElements in componentDidUpdate", async () => {
      const page = await newSpecPage({
        components: [Drawer],
        html: `<ic-drawer heading="Menu" message="test" expanded="true"></ic-drawer>`,
      });

      const updateInteractiveElementsSpy = jest.spyOn(
        page.rootInstance,
        "updateInteractiveElements"
      );

      page.root.expanded = !page.root.expanded;

      page.rootInstance.componentDidUpdate();
      await page.waitForChanges();

      expect(updateInteractiveElementsSpy).toHaveBeenCalledTimes(2);
    });
  });
});
