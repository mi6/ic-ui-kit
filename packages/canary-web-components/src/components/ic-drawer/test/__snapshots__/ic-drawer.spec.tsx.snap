// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-drawer should render at the bottom of the viewport 1`] = `
<ic-drawer class="collapsed ic-bottom-position ic-drawer" heading="Menu" message="test" position="bottom">
  <mock:shadow-root>
    <div class="collapsed-drawer">
      <ic-button aria-label="Open drawer" class="chevron-btn" size="small" variant="icon">
        svg
      </ic-button>
    </div>
    <div aria-labelledby="drawer-heading" class="ic-drawer-panel medium"></div>
  </mock:shadow-root>
</ic-drawer>
`;

exports[`ic-drawer should render at the top of the viewport 1`] = `
<ic-drawer class="collapsed ic-drawer ic-top-position" heading="Menu" message="test" position="top">
  <mock:shadow-root>
    <div class="collapsed-drawer">
      <ic-button aria-label="Open drawer" class="chevron-btn" size="small" variant="icon">
        svg
      </ic-button>
    </div>
    <div aria-labelledby="drawer-heading" class="ic-drawer-panel medium"></div>
  </mock:shadow-root>
</ic-drawer>
`;

exports[`ic-drawer should render on the left 1`] = `
<ic-drawer class="collapsed ic-drawer ic-left-position" heading="Menu" message="test" position="left">
  <mock:shadow-root>
    <div class="collapsed-drawer">
      <ic-button aria-label="Open drawer" class="chevron-btn" size="small" variant="icon">
        svg
      </ic-button>
    </div>
    <div aria-labelledby="drawer-heading" class="ic-drawer-panel medium"></div>
  </mock:shadow-root>
</ic-drawer>
`;

exports[`ic-drawer should render on the right 1`] = `
<ic-drawer class="collapsed ic-drawer ic-right-position" heading="Menu" message="test" position="right">
  <mock:shadow-root>
    <div class="collapsed-drawer">
      <ic-button aria-label="Open drawer" class="chevron-btn" size="small" variant="icon">
        svg
      </ic-button>
    </div>
    <div aria-labelledby="drawer-heading" class="ic-drawer-panel medium"></div>
  </mock:shadow-root>
</ic-drawer>
`;
